Add utility to get computed kernel in torch.library

mikaylagawarecki opened the pull request:
Adds `OperatorEntry::getComputedKernelForDispatchKey` which returns the KernelFunction corresponding to `OperatorEntry.dispatchTable_[dispatch_ix]` for a given dispatch key
- Specifically it returns a `SafeKernelFunction` that holds a `KernelToken`. This `KernelToken` is registered to the `KernelFunction` in `OperatorEntry.kernels_` and will be invalidated when the `KernelFunction` is destructed (i.e. when the `AnnotatedKernel` that holds this `KernelFunction` is removed from `kernels_`, which happens when the corresponding impl is deregistered).
- `SafeKernelFunction` can be called via `callBoxed`, the validity of the token will be checked before this happens
- `SafeKernelFunction` is pybinded and `getComputedKernelForDispatchKey` is exposed to the frontend ia `torch.library.get_kernel`

Related to https://github.com/pytorch/pytorch/issues/155330

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158393



cc @albanD

mikaylagawarecki committed:
Add utility to get computed kernel in torch.library

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Related to https://github.com/pytorch/pytorch/issues/155330




[ghstack-poisoned]

mikaylagawarecki commented on test/test_custom_ops.py:
```diff
@@ -4147,6 +4148,131 @@ def test_any_output_is_alias_to_input_or_output(self):
+
+            kernel = torch.library.get_kernel("aten::arange.start", "CPU")
+            op_handle = torch.ops.aten.arange.start._handle
+            dispatch_keys = torch._C.DispatchKeySet(torch._C.DispatchKey.CPU)
```
Something I'm not too sure about is how we expect users of this API to construct the DispatchKeySet to pass to `kernel.call_boxed`

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Related to https://github.com/pytorch/pytorch/issues/155330




[ghstack-poisoned]

mikaylagawarecki commented on aten/src/ATen/core/boxing/KernelFunction.h:
```diff
@@ -90,6 +95,12 @@ class TORCH_API KernelFunction final {
+  KernelFunction(const KernelFunction&) = default;
+  KernelFunction& operator=(const KernelFunction&) = default;
+
+  KernelFunction(KernelFunction&&) noexcept = default;
```
testing whether this line fixes 

```
␛[K/var/lib/jenkins/workspace/torch/library.h:840:17:␛[K ␛[Kerror: ␛[Kuse of deleted function ‘␛[Ktorch::CppFunction::CppFunction(torch::CppFunction&&)␛[K’
2025-07-16T20:31:39.5168847Z   840 |     CppFunction ␛[Kf␛[K((std::forward<Func>(raw_f)));
2025-07-16T20:31:39.5169347Z       |                 ␛[K^␛[K
2025-07-16T20:31:39.5171059Z ␛[K/var/lib/jenkins/workspace/torch/library.h:212:3:␛[K ␛[Knote: ␛[K‘␛[Ktorch::CppFunction::CppFunction(torch::CppFunction&&) noexcept␛[K’ is implicitly deleted because its exception-specification does not match the implicit exception-specification 
```

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Related to https://github.com/pytorch/pytorch/issues/155330




cc albanD

[ghstack-poisoned]

albanD reviewed (commented):
Sounds pretty good!
Only small questions!

mikaylagawarecki commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -315,6 +315,29 @@ const AnnotatedKernel* OperatorEntry::getKernelForDispatchKey(DispatchKey dispat
+  auto [annotatedKernel, _] =
+      computeDispatchTableEntryWithDebug(c10::Dispatcher::singleton(), k);
+
+  return SafeKernelFunction(&annotatedKernel.kernel);
```
This gives something like 

```
SafeKernelFunction(debug='registered at /data/users/mg1998/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd_0.cpp:2309')
```

Do you think that is meaningful enough?

albanD replied:
Yes, it is 100% super useful. These error messages saved me a few times for multiple-registration errors!
And for python users, it should point to their code directly. Which is even better so they know which function this is!

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Adds `OperatorEntry::getComputedKernelForDispatchKey` which returns the KernelFunction corresponding to `OperatorEntry.dispatchTable_[dispatch_ix]` for a given dispatch key
- Specifically it returns a `SafeKernelFunction` that holds a `KernelToken`. This `KernelToken` is registered to the `KernelFunction` in `OperatorEntry.kernels_` and will be invalidated when the `KernelFunction` is destructed (i.e. when the `AnnotatedKernel` that holds this `KernelFunction` is removed from `kernels_`, which happens when the corresponding impl is deregistered).
- `SafeKernelFunction` can be called via `callBoxed`, the validity of the token will be checked before this happens
- `SafeKernelFunction` is pybinded and `getComputedKernelForDispatchKey` is exposed to the frontend ia `torch.library.get_kernel`

Related to https://github.com/pytorch/pytorch/issues/155330




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Adds `OperatorEntry::getComputedKernelForDispatchKey` which returns the KernelFunction corresponding to `OperatorEntry.dispatchTable_[dispatch_ix]` for a given dispatch key
- Specifically it returns a `SafeKernelFunction` that holds a `KernelToken`. This `KernelToken` is registered to the `KernelFunction` in `OperatorEntry.kernels_` and will be invalidated when the `KernelFunction` is destructed (i.e. when the `AnnotatedKernel` that holds this `KernelFunction` is removed from `kernels_`, which happens when the corresponding impl is deregistered).
- `SafeKernelFunction` can be called via `callBoxed`, the validity of the token will be checked before this happens
- `SafeKernelFunction` is pybinded and `getComputedKernelForDispatchKey` is exposed to the frontend ia `torch.library.get_kernel`

Related to https://github.com/pytorch/pytorch/issues/155330




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Adds `OperatorEntry::getComputedKernelForDispatchKey` which returns the KernelFunction corresponding to `OperatorEntry.dispatchTable_[dispatch_ix]` for a given dispatch key
- Specifically it returns a `SafeKernelFunction` that holds a `KernelToken`. This `KernelToken` is registered to the `KernelFunction` in `OperatorEntry.kernels_` and will be invalidated when the `KernelFunction` is destructed (i.e. when the `AnnotatedKernel` that holds this `KernelFunction` is removed from `kernels_`, which happens when the corresponding impl is deregistered).
- `SafeKernelFunction` can be called via `callBoxed`, the validity of the token will be checked before this happens
- `SafeKernelFunction` is pybinded and `getComputedKernelForDispatchKey` is exposed to the frontend ia `torch.library.get_kernel`

Related to https://github.com/pytorch/pytorch/issues/155330




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Adds `OperatorEntry::getComputedKernelForDispatchKey` which returns the KernelFunction corresponding to `OperatorEntry.dispatchTable_[dispatch_ix]` for a given dispatch key
- Specifically it returns a `SafeKernelFunction` that holds a `KernelToken`. This `KernelToken` is registered to the `KernelFunction` in `OperatorEntry.kernels_` and will be invalidated when the `KernelFunction` is destructed (i.e. when the `AnnotatedKernel` that holds this `KernelFunction` is removed from `kernels_`, which happens when the corresponding impl is deregistered).
- `SafeKernelFunction` can be called via `callBoxed`, the validity of the token will be checked before this happens
- `SafeKernelFunction` is pybinded and `getComputedKernelForDispatchKey` is exposed to the frontend ia `torch.library.get_kernel`

Related to https://github.com/pytorch/pytorch/issues/155330




cc albanD

[ghstack-poisoned]

albanD commented on torch/library.py:
```diff
@@ -1475,6 +1476,84 @@ def get_ctx() -> "torch._library.fake_impl.FakeImplCtx":
+        >>> kernel = torch.library.get_kernel(torch.ops.aten.add.Tensor, "CPU")
+        >>>
+        >>> # Use the kernel to call the operator
+        >>> op_handle = torch._C._dispatch_find_schema_or_throw("aten::add", "Tensor")
```
Can we store the handle on the object returned by get_kernel so the user doesn't have to do this by any chance?

albanD commented on torch/library.py:
```diff
@@ -1475,6 +1476,84 @@ def get_ctx() -> "torch._library.fake_impl.FakeImplCtx":
+        >>> original_sin_kernel = torch.library.get_kernel("aten::sin", "CPU")
+        >>>
+        >>> def conditional_sin_impl(dispatch_keys, x):
+        >>> # If input has negative values, use original sin, otherwise return zeros
```
wrong identation

albanD commented on torch/library.py:
```diff
@@ -1475,6 +1476,84 @@ def get_ctx() -> "torch._library.fake_impl.FakeImplCtx":
+        >>> op_handle = torch._C._dispatch_find_schema_or_throw("aten::add", "Tensor")
+        >>> a = torch.tensor([1.0, 2.0])
+        >>> b = torch.tensor([3.0, 4.0])
+        >>> result = kernel.call_boxed(op_handle, torch._C.DispatchKeySet("CPU"), a, b)
```
Why do we have a keyset here? Isn't this supposed to be just a kernel? We're not really going through full dispatch here right?

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Adds `OperatorEntry::getComputedKernelForDispatchKey` which returns the KernelFunction corresponding to `OperatorEntry.dispatchTable_[dispatch_ix]` for a given dispatch key
- Specifically it returns a `SafeKernelFunction` that holds a `KernelToken`. This `KernelToken` is registered to the `KernelFunction` in `OperatorEntry.kernels_` and will be invalidated when the `KernelFunction` is destructed (i.e. when the `AnnotatedKernel` that holds this `KernelFunction` is removed from `kernels_`, which happens when the corresponding impl is deregistered).
- `SafeKernelFunction` can be called via `callBoxed`, the validity of the token will be checked before this happens
- `SafeKernelFunction` is pybinded and `getComputedKernelForDispatchKey` is exposed to the frontend ia `torch.library.get_kernel`

Related to https://github.com/pytorch/pytorch/issues/155330




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Adds `OperatorEntry::getComputedKernelForDispatchKey` which returns the KernelFunction corresponding to `OperatorEntry.dispatchTable_[dispatch_ix]` for a given dispatch key
- Specifically it returns a `SafeKernelFunction` that holds a `KernelToken`. This `KernelToken` is registered to the `KernelFunction` in `OperatorEntry.kernels_` and will be invalidated when the `KernelFunction` is destructed (i.e. when the `AnnotatedKernel` that holds this `KernelFunction` is removed from `kernels_`, which happens when the corresponding impl is deregistered).
- `SafeKernelFunction` can be called via `callBoxed`, the validity of the token will be checked before this happens
- `SafeKernelFunction` is pybinded and `getComputedKernelForDispatchKey` is exposed to the frontend ia `torch.library.get_kernel`

Related to https://github.com/pytorch/pytorch/issues/155330




cc albanD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get computed kernel in torch.library"

Adds `OperatorEntry::getComputedKernelForDispatchKey` which returns the KernelFunction corresponding to `OperatorEntry.dispatchTable_[dispatch_ix]` for a given dispatch key
- Specifically it returns a `SafeKernelFunction` that holds a `KernelToken`. This `KernelToken` is registered to the `KernelFunction` in `OperatorEntry.kernels_` and will be invalidated when the `KernelFunction` is destructed (i.e. when the `AnnotatedKernel` that holds this `KernelFunction` is removed from `kernels_`, which happens when the corresponding impl is deregistered).
- `SafeKernelFunction` can be called via `callBoxed`, the validity of the token will be checked before this happens
- `SafeKernelFunction` is pybinded and `getComputedKernelForDispatchKey` is exposed to the frontend ia `torch.library.get_kernel`

Related to https://github.com/pytorch/pytorch/issues/155330




cc albanD

[ghstack-poisoned]

albanD reviewed (approved):
Awesome!
Thanks!

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
